### PR TITLE
rewrote manage_case in python

### PR DESCRIPTION
**Pull request by mvertens**
Rewrote manage_case in python. This required also adding the following new files that primarily are being used to write the output of the query commands.
utils/python/CIME/XML/compsets.py
utils/python/CIME/XML/grids.py

### CODE REVIEW

File: scripts/manage_case [hunk 1]
```diff
@@ -1,375 +1,172 @@
...
-    my $xml_machine = XML::LibXML->new( no_blanks => 1)->parse_file($file);
-    foreach my $machine ($xml_machine->findnodes(".//machine")) {
-	my $name = $machine->getAttribute('MACH');
-	foreach my $node ($machine->findnodes("./DESC")) {
-	    my $desc = $node->textContent();
-	    print " $name :  \n      $desc \n";
-	}
-    }
-}
+#!/usr/bin/env python2.7
```
Comment (jedwards4b) [inline scripts/manage_case]: I think that you just want python2 here not 2.7

File: scripts/manage_case [hunk 2]
```diff
@@ -1,375 +1,172 @@
...
+from CIME.XML.machines  import Machines
+
+import argparse, doctest
+
+###############################################################################
+def query_grids():  
+###############################################################################
+
+    files = Files()
+    config_file = files.get_value("GRIDS_SPEC_FILE", resolved=True)
```
Comment (jedwards4b) [inline scripts/manage_case]: resolved defaults to True so no need to have that argument here (it doesn't hurt anything to have it either)

File: scripts/manage_case [hunk 3]
```diff
@@ -1,375 +1,172 @@
...
+
+###############################################################################
+def query_compsets(name):  
+###############################################################################
+
+    # Determine valid component values by checking the value attributes for COMPSETS_SPEC_FILE
+    files = Files()
+    components = files.get_components("COMPSETS_SPEC_FILE", resolved=True)
+    match_found = "unset"
+    for component in components:
```
Comment (jedwards4b) [inline scripts/manage_case]: I think that this can be reduced to 
  if name in components:
       match_found = name

File: scripts/manage_case (hunk 1 above)
Comment (mvertens) [inline scripts/manage_case]: Yes - that is why the comment. I could not get it to work on my mac without this. Should I put in another pull request with this fix?

File: scripts/manage_case [hunk 4]
```diff
@@ -1,375 +1,172 @@
...
+    expect((config_file),
+           "Cannot find any config_component.xml file for %s" %name)
+
+    # Check that file exists on disk
+    expect(os.path.isfile(config_file),
+           "Cannot find config_file %s on disk" %config_file)
+
+    # Now parse the compsets file and write out the compset alias and longname as well as the help text
+    # determine component xml content 
+    parsed_xml = Compsets(config_file)
```
Comment (jedwards4b) [inline scripts/manage_case]: I think that you shouldn't resuse the same name (parsed_xml) for different xml objects - call each object what it is
(compset, machine, etc)

File: utils/python/CIME/XML/component.py [hunk 5]
```diff
@@ -22,7 +22,31 @@ def get_value(self, name, attribute={}, resolved=False):
...
+            rootnode = self.get_node("help")
+            helptext = rootnode.text
+            return helptext
+        elif name == "description":
+            rootnode = self.get_node("description")
+            compsets = {}
+            descs = self.get_nodes("desc", root=rootnode)
+            for desc in descs:
+                attrib = desc.attrib["compset"]
+                text = desc.text
```
Comment (jedwards4b) [inline utils/python/CIME/XML/component.py]: why not
compsets[desc.attrib["compset"]] = desc.text

File: utils/python/CIME/XML/files.py [hunk 6]
```diff
@@ -20,3 +20,12 @@ def __init__(self):
         """
         infile = os.path.join(get_cime_root(), "cime_config", get_model(), "config_files.xml")
         EntryID.__init__(self, infile)
+
+    def get_components(self, nodename, resolved=True):        
+        node = self.get_optional_node("entry", {"id":nodename})
+        valnodes = self.get_nodes("value", root=node)
+        values = list();
+        for valnode in valnodes:
+            value = valnode.attrib["component"]
```
Comment (jedwards4b) [inline utils/python/CIME/XML/files.py]: values.append(valnode.attrib["component"])

File: utils/python/CIME/XML/grids.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+                domain_names = gridRE.split(lname.text)[1:]
+                domain_set = list(set(domain_names))
+                for domain in domain_set:
+                    print "   domain is ",domain
+                    domain_node = self.get_node(nodename="domain", attributes={"name":domain}) 
+                    for child in domain_node:
+                        print  "     ",child.tag,": ",child.text
+                    
+                # in verbose mode add grid maps 
+                # FIXME (mvertens, 3/2016) - how do I make the following more consise and less hard-wired
```
Comment (jedwards4b) [inline utils/python/CIME/XML/grids.py]: You should get the components from driver_cpl/cime_config/config_component.xml
using  component.get_value("component")  then create dicts like grid["atm"], grid["lnd"] etc

**Comment by jedwards4b** [comment]
Review complete.

File: scripts/manage_case (hunk 1 above)
Comment (jgfouca) [inline scripts/manage_case]: @mvertens , no that's way too much work. Just make additional commits to your branch and push.

Comment (jedwards4b) [inline scripts/manage_case]: You can just update this PR by pushing to your fork.

File: scripts/manage_case [hunk 8]
```diff
@@ -1,375 +1,172 @@
...
+def query_grids():  
+###############################################################################
+
+    files = Files()
+    config_file = files.get_value("GRIDS_SPEC_FILE", resolved=True)
+    expect(os.path.isfile(config_file),
+           "Cannot find config_file %s on disk" %config_file)
+
+    parsed_xml = Grids(config_file)
+    if verbose:
```
Comment (jgfouca) [inline scripts/manage_case]: You should probably use the logger stuff that Jim set up. Also, where does verbose come from in this function? I don't see it as an argument or stack variable. You may want to run pylint on this code to find mistakes like this.

File: scripts/manage_case [hunk 9]
```diff
@@ -1,375 +1,172 @@
...
+def query_machines():  
+###############################################################################
+
+    files = Files()
+    config_file = files.get_value("MACHINES_SPEC_FILE", resolved=True)
+    expect(os.path.isfile(config_file),
+           "Cannot find config_file %s on disk" %config_file)
+
+    parsed_xml = Machines(config_file)
+    if verbose:
```
Comment (jgfouca) [inline scripts/manage_case]: See prior comment.

File: scripts/manage_case [hunk 10]
```diff
@@ -1,375 +1,172 @@
...
+    comps_array = comps_object.get_value("component")
+    
+    # Loop through the elements for each component class (in config_files.xml) 
+    # and see if there is a match for the the target component in the component attribute 
+    match_found = False
+    valid_components = list()
+    for comp in comps_array:
+        string = "CONFIG_%s_FILE"%comp 
+
+        # determine all components in string
```
Comment (jgfouca) [inline scripts/manage_case]: As Jim said, please get rid of all these resolve=True.

File: scripts/manage_case [hunk 11]
```diff
@@ -1,375 +1,172 @@
...
+        parsed_xml.print_values()
+
+###############################################################################
+def query_compsets(name):  
+###############################################################################
+
+    # Determine valid component values by checking the value attributes for COMPSETS_SPEC_FILE
+    files = Files()
+    components = files.get_components("COMPSETS_SPEC_FILE", resolved=True)
+    match_found = "unset"
```
Comment (jgfouca) [inline scripts/manage_case]: "unset" is a very non-pythonic choice to indicate not found. Use python's None instead.

File: scripts/manage_case [hunk 12]
```diff
@@ -1,375 +1,172 @@
...
+        # determine if config_file is on disk
+        config_file = files.get_value(string, attribute={"component":name}, resolved=True)
+        if config_file is not None:
+            match_found = True
+            expect(os.path.isfile(config_file),
+                   "Cannot find config_file %s on disk" %config_file)
+            break
+
+    # If name is not a valid argument - exit with error
+    expect((match_found),
```
Comment (jgfouca) [inline scripts/manage_case]: No need for parens around match_found.

File: scripts/manage_case [hunk 13]
```diff
@@ -1,375 +1,172 @@
...
+def parse_command_line(args):
+###############################################################################
+
+    cime_model = CIME.utils.get_model()
+
+    parser = argparse.ArgumentParser()
+
+    CIME.utils.setup_standard_logging_options(parser)
+
+    parser.add_argument("--query_compsets_setby",
```
Comment (jgfouca) [inline scripts/manage_case]: It is not POSIX compliant to have --long_options_like_this. You want --long-options-like-this.

File: scripts/manage_case [hunk 14]
```diff
@@ -1,375 +1,172 @@
...
+
+    args = parser.parse_args()
+
+    CIME.utils.handle_standard_logging_options(args)
+
+    return args
+
+###############################################################################
+
+args = parse_command_line(sys.argv)
```
Comment (jgfouca) [inline scripts/manage_case]: It is generally not a good idea to do work in the global namespace. Please follow the convention we've used in our other scripts:
`def _main_func():
    *do stuff*

if __name__ == "__main__":
    _main_func()`

File: utils/python/CIME/XML/component.py [hunk 15]
```diff
@@ -22,7 +22,31 @@ def get_value(self, name, attribute={}, resolved=False):
             compnode = self.get_node("components")
```
Comment (jgfouca) [inline utils/python/CIME/XML/component.py]: This function confuses me and could use a docstring to explain. The return value can be a string, list, or dictionary depending on name? I'd rather have list, dictionary etc as a <type> than having special cases everywhere.

File: utils/python/CIME/XML/grids.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+            gridmap_nodes = self.get_nodes(nodename="gridmap")
+            gridRE = re.compile(r"[_]{0,1}[a-z]{1,2}%")
+
+        # write out grid elements 
+        grid_nodes = self.get_nodes(nodename="grid")
+        for grid_node in grid_nodes:
+            lname   = grid_node.find("lname")
+            sname   = grid_node.find("sname")
+            support = grid_node.find("support")
+            alias   = grid_node.find("alias")
```
Comment (jgfouca) [inline utils/python/CIME/XML/grids.py]: @jedwards4b can correct me if I'm wrong, but I thought we wanted the logger to handle output and debug outout?

File: utils/python/CIME/XML/grids.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+                # in verbose mode add grid maps 
+                # FIXME (mvertens, 3/2016) - how do I make the following more consise and less hard-wired 
+                atm_grid  = domain_names[0]
+                lnd_grid  = domain_names[1]
+                ocn_grid  = domain_names[2]
+                rof_grid  = domain_names[3]
+                mask_grid = domain_names[4]
+                glc_grid  = domain_names[5]
+                wav_grid  = domain_names[6]
+
```
Comment (jgfouca) [inline utils/python/CIME/XML/grids.py]: grids = [ ("atm_grid", atm_grid), "lnd_grid", lnd_grid), ...]
    for idx, grid in enumerate(grids):
        for other_grid in grids[idx+1:]:
            nodes = self.get_nodes(nodename="gridmap", attributes={grid[0]:grid[1], other_grid[0]:other_grid[1]})
            for gridmap_node in nodes:
                for child in gridmap_node:
                    print " ", child.tag, ": ", child.text

**Comment by jgfouca** [comment]
Review complete.

File: utils/python/CIME/XML/component.py (hunk 5 above)
Comment (mvertens) [inline utils/python/CIME/XML/component.py]: could you please clarify in the example above what you mean...

File: utils/python/CIME/XML/files.py (hunk 6 above)
Comment (mvertens) [inline utils/python/CIME/XML/files.py]: could you also clarify in the example above what you mean - it would be really helpful

File: scripts/manage_case (hunk 10 above)
Comment (mvertens) [inline scripts/manage_case]: Done

File: scripts/manage_case (hunk 12 above)
Comment (mvertens) [inline scripts/manage_case]: Done

File: scripts/manage_case (hunk 13 above)
Comment (mvertens) [inline scripts/manage_case]: Done

File: scripts/manage_case (hunk 14 above)
Comment (mvertens) [inline scripts/manage_case]: Done

File: utils/python/CIME/XML/component.py (hunk 5 above)
Comment (mvertens) [inline utils/python/CIME/XML/component.py]: I think my may is clearer - so I'll keep it that way for now. We agreed on this.
Done

**Comment by jgfouca** [comment]
@mvertens , is this ready?

**Comment by mvertens** [comment]
Yes - I pushed this to my branch that I had issued a pull request for - and
I thought this would automatically update you and Jim.

**Comment by jgfouca** [comment]
@mvertens Looks much better! Looks like you addressed every comment.
